File: demos/csharp/AlgorithmsDemo.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace AlgorithmsTopic;

public sealed record Score(string Name, int Value);

public static class Program
{
    private static IReadOnlyList<string> Bfs(IReadOnlyDictionary<string, string[]> graph, string start)
    {
        var order = new List<string>();
        var seen = new HashSet<string>();
        var queue = new Queue<string>();
        queue.Enqueue(start);
        while (queue.TryDequeue(out var node))
        {
            if (!seen.Add(node)) continue;
            order.Add(node);
            if (graph.TryGetValue(node, out var next))
                foreach (var value in next) queue.Enqueue(value);
        }
        return order;
    }

    public static void Main()
    {
        var ranked = new[] { new Score("Lin", 91), new Score("Ada", 95), new Score("Kai", 91) }
            .OrderByDescending(item => item.Value).ThenBy(item => item.Name);
        var numbers = new List<int> { 2, 5, 8, 13, 21 };
        var graph = new Dictionary<string, string[]> { ["A"] = ["B", "C"], ["B"] = ["D"], ["C"] = ["D"] };

        Console.WriteLine("sorted=" + string.Join(',', ranked.Select(item => $"{item.Name}:{item.Value}")));
        Console.WriteLine($"binary-search-13={numbers.BinarySearch(13)}");
        Console.WriteLine("bfs=" + string.Join(',', Bfs(graph, "A")));
    }
}

File: demos/python/algorithms_demo.py
from bisect import bisect_left
from collections import deque


def bfs(graph: dict[str, list[str]], start: str) -> list[str]:
    order: list[str] = []
    seen: set[str] = set()
    queue = deque([start])
    while queue:
        node = queue.popleft()
        if node in seen:
            continue
        seen.add(node)
        order.append(node)
        queue.extend(graph.get(node, []))
    return order


scores = [("Lin", 91), ("Ada", 95), ("Kai", 91)]
ranked = sorted(scores, key=lambda item: (-item[1], item[0]))
numbers = [2, 5, 8, 13, 21]
graph = {"A": ["B", "C"], "B": ["D"], "C": ["D"]}

print(f"sorted={ranked}")
print(f"binary-search-13={bisect_left(numbers, 13)}")
print(f"bfs={bfs(graph, 'A')}")
